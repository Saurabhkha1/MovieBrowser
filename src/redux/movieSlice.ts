import {createSlice, createAsyncThunk, PayloadAction} from '@reduxjs/toolkit';
import {fetchMovies} from '../network';

export type Movie = {
  id: number;
  title: string;
  poster_path: string;
  release_date: string;
  vote_average: number;
  overview: string;
};

export type MoviesState = {
  nowPlaying: Movie[];
  popular: Movie[];
  topRated: Movie[];
  upcoming: Movie[];
  status: 'idle' | 'loading' | 'succeeded' | 'failed';
  error: string | null;
  page: number;
  hasMore: boolean;
};

const initialState: MoviesState = {
  nowPlaying: [],
  popular: [],
  topRated: [],
  upcoming: [],
  status: 'idle',
  error: null,
  page: 1,
  hasMore: true,
};

export const fetchMoviesByCategory = createAsyncThunk(
  'movies/fetchMoviesByCategory',
  async ({
    category,
    page,
  }: {
    category: 'now_playing' | 'popular' | 'top_rated' | 'upcoming';
    page: number;
  }) => {
    const movies = await fetchMovies(category, page);
    return {category, movies, page};
  },
);

const moviesSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    resetState: state => {
      state.nowPlaying = [];
      state.popular = [];
      state.topRated = [];
      state.upcoming = [];
      state.status = 'idle';
      state.error = null;
      state.page = 1;
      state.hasMore = true;
    },
  },
  extraReducers: builder => {
    builder
      .addCase(fetchMoviesByCategory.pending, state => {
        state.status = 'loading';
        state.error = null;
      })
      .addCase(
        fetchMoviesByCategory.fulfilled,
        (
          state,
          action: PayloadAction<{
            category: string;
            movies: Movie[];
            page: number;
          }>,
        ) => {
          state.status = 'succeeded';
          if (action.payload.movies.length === 0) {
            state.hasMore = false;
          }
          switch (action.payload.category) {
            case 'now_playing':
              if (action.payload.page === 1) {
                state.nowPlaying = action.payload.movies;
              } else {
                state.nowPlaying = [
                  ...state.nowPlaying,
                  ...action.payload.movies,
                ];
              }
              break;
            case 'popular':
              if (action.payload.page === 1) {
                state.popular = action.payload.movies;
              } else {
                state.popular = [...state.popular, ...action.payload.movies];
              }
              break;
            case 'top_rated':
              if (action.payload.page === 1) {
                state.topRated = action.payload.movies;
              } else {
                state.topRated = [...state.topRated, ...action.payload.movies];
              }
              break;
            case 'upcoming':
              if (action.payload.page === 1) {
                state.upcoming = action.payload.movies;
              } else {
                state.upcoming = [...state.upcoming, ...action.payload.movies];
              }
              break;
            default:
              break;
          }
          state.page = action.payload.page;
        },
      )
      .addCase(fetchMoviesByCategory.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message || 'Failed to fetch movies';
      });
  },
});

export const {resetState} = moviesSlice.actions;
export default moviesSlice.reducer;
