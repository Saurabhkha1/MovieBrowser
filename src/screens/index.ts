export * from './commonHome';
export * from './movieDetails';
