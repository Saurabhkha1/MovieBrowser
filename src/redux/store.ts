import {configureStore} from '@reduxjs/toolkit';
import movieSlice from './movieSlice';

export const store = configureStore({
  reducer: {movieReducer: movieSlice},
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
