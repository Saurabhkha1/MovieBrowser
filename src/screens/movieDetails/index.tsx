import React from 'react';
import {
  View,
  Text,
  Image,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
} from 'react-native';
import {Movie} from '../../redux/movieSlice';
import {Colors} from '../../utils/colors';
import {IMAGE_BASE_URL} from '../../utils/constants';
import {moderateScale} from '../../utils/helper';
import {IconComponent} from '../../components';
import {Icons} from '../../utils/icons';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../navigation';

type Props = {
  route: {
    params: {
      movie: Movie;
    };
  };
};

export const MovieDetails = ({route}: Props) => {
  const {movie} = route?.params;
  const navigation =
    useNavigation<NativeStackNavigationProp<RootStackParamList>>();
  return (
    <>
      <TouchableOpacity
        style={styles.backIcon}
        onPress={() => navigation.pop()}>
        <IconComponent icon={Icons.backIcon} size={moderateScale(25)} />
      </TouchableOpacity>
      <ScrollView style={styles.container}>
        <Image
          source={{uri: IMAGE_BASE_URL + movie.poster_path}}
          style={styles.image}
          resizeMode="cover"
        />

        <View style={styles.detailsContainer}>
          <Text style={styles.title}>{movie.title}</Text>
          <Text style={styles.subtitle}>{movie.release_date}</Text>
          <Text style={styles.description}>{movie.overview}</Text>
        </View>
      </ScrollView>
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.white,
  },
  image: {
    width: '100%',
    height: moderateScale(500),
    borderBottomWidth: 1,
    borderBottomColor: Colors.grey,
  },
  detailsContainer: {
    padding: moderateScale(15),
  },
  title: {
    fontSize: moderateScale(24),
    fontWeight: 'bold',
    marginBottom: moderateScale(10),
  },
  subtitle: {
    fontSize: moderateScale(18),
    color: Colors.darkGrey,
    marginBottom: moderateScale(15),
  },
  description: {
    fontSize: moderateScale(16),
    color: Colors.black,
    lineHeight: moderateScale(24),
  },
  backIcon: {
    position: 'absolute',
    width: moderateScale(40),
    height: moderateScale(40),
    backgroundColor: Colors.grey,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: moderateScale(30),
    top: moderateScale(5),
    left: moderateScale(5),
    zIndex: 1,
  },
});
