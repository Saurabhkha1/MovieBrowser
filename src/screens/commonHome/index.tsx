import {
  View,
  StyleSheet,
  FlatList,
  Text,
  ActivityIndicator,
} from 'react-native';
import React, {useEffect, useMemo, useRef, useState} from 'react';
import {Header, Loader, MovieComponent} from '../../components';
import {IconProp} from '@fortawesome/fontawesome-svg-core';
import {Colors} from '../../utils/colors';
import {moderateScale} from '../../utils/helper';
import {AppDispatch, RootState} from '../../redux/store';
import {useDispatch, useSelector} from 'react-redux';
import {fetchMoviesByCategory, Movie, resetState} from '../../redux/movieSlice';

type Props = {
  route: {
    params: {
      title: string;
      icon: IconProp;
    };
  };
};

const FooterComponent = ({loading}: {loading: boolean}) => (
  <View style={styles.footer}>
    {loading && <ActivityIndicator size="large" color={Colors.black} />}
  </View>
);

export const CommonHome = ({route}: Props) => {
  const {title, icon} = route?.params;
  const dispatch = useDispatch<AppDispatch>();
  const {
    nowPlaying,
    popular,
    topRated,
    upcoming,
    status,
    error,
    page,
    hasMore,
  } = useSelector((state: RootState) => state.movieReducer);
  const [isFetching, setIsFetching] = useState(false);
  const category = useRef<'now_playing' | 'popular' | 'top_rated' | 'upcoming'>(
    'now_playing',
  );
  const [searchValue, setSearchValue] = useState('');
  const [searchedMovies, setSearchedMovies] = useState<Movie[]>();

  const pageRef = useRef(1);

  useEffect(() => {
    pageRef.current = page;
  }, [page]);

  useEffect(() => {
    dispatch(resetState());
  }, [title, dispatch]);

  useEffect(() => {
    if (status === 'idle') {
      switch (title.toLowerCase()) {
        case 'now playing':
          category.current = 'now_playing';
          break;
        case 'popular':
          category.current = 'popular';
          break;
        case 'top rated':
          category.current = 'top_rated';
          break;
        case 'upcoming':
          category.current = 'upcoming';
          break;
        default:
          return;
      }
      dispatch(fetchMoviesByCategory({category: category.current, page: 1}));
    }
  }, [dispatch, title, status]);

  let movies: Movie[] = useMemo(() => [], []);
  switch (title.toLowerCase()) {
    case 'now playing':
      movies = nowPlaying;
      break;
    case 'popular':
      movies = popular;
      break;
    case 'top rated':
      movies = topRated;
      break;
    case 'upcoming':
      movies = upcoming;
      break;
    default:
      break;
  }

  const loadMoreMovies = () => {
    if (hasMore && !isFetching && status !== 'loading') {
      setIsFetching(true);
      pageRef.current += 1;
      dispatch(
        fetchMoviesByCategory({
          category: category.current,
          page: page + 1,
        }),
      ).then(() => {
        setIsFetching(false);
      });
    }
  };

  useEffect(() => {
    if (searchValue?.length) {
      let timeout = setTimeout(() => {
        const searchedMoviesTemp: Movie[] = movies.filter(item =>
          item?.title?.toLowerCase()?.includes(searchValue?.toLowerCase()),
        );
        setSearchedMovies(searchedMoviesTemp);
      }, 500);
      return () => clearTimeout(timeout);
    }
  }, [searchValue, movies]);

  return (
    <View style={styles.container}>
      <Header
        title={title}
        icon={icon}
        searchValue={searchValue}
        onChangeText={text => setSearchValue(text)}
      />
      <View style={styles.mainView}>
        {status === 'loading' && pageRef.current === 1 && <Loader />}
        {status === 'failed' && <Text>Error: {error}</Text>}
        {(status === 'succeeded' || isFetching) && (
          <FlatList
            style={styles.flatlistStyle}
            data={searchValue?.length ? searchedMovies : movies}
            keyExtractor={item => item.id.toString()}
            numColumns={2}
            renderItem={({item}) => <MovieComponent movie={item} />}
            ListFooterComponent={<FooterComponent loading={isFetching} />}
            onEndReached={searchValue?.length ? () => null : loadMoreMovies}
            onEndReachedThreshold={0.5}
          />
        )}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.white,
  },
  mainView: {
    padding: moderateScale(10),
  },
  footer: {
    paddingVertical: moderateScale(20),
    alignItems: 'center',
  },
  flatlistStyle: {height: '100%'},
});
