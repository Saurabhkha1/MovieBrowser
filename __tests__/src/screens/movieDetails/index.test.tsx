import React from 'react';
import renderer from 'react-test-renderer';
import {useNavigation} from '@react-navigation/native';
import {Movie} from '../../../../src/redux/movieSlice';
import {MovieDetails} from '../../../../src/screens/MovieDetails';
import {TouchableOpacity} from 'react-native';

describe('MovieDetails', () => {
  const mockMovie: Movie = {
    id: 1,
    title: 'Test Movie',
    poster_path: '/test-poster.jpg',
    release_date: '2023-01-01',
    overview: 'Test movie overview',
    vote_average: 1.5,
  };

  const route = {
    params: {
      movie: mockMovie,
    },
  };

  const navigation = {
    pop: jest.fn(),
  };

  beforeEach(() => {
    (useNavigation as jest.Mock).mockReturnValue(navigation);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<MovieDetails route={route} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls navigation.pop on back button press', () => {
    const component = renderer.create(<MovieDetails route={route} />);
    const backButton = component.root.findByType(TouchableOpacity);

    renderer.act(() => {
      backButton.props.onPress();
    });

    expect(navigation.pop).toHaveBeenCalled();
  });

  it('displays the correct movie details', () => {
    const component = renderer.create(<MovieDetails route={route} />);
    const instance = component.root;

    const title = instance.findByProps({children: mockMovie.title}).props
      .children;
    const releaseDate = instance.findByProps({children: mockMovie.release_date})
      .props.children;
    const overview = instance.findByProps({children: mockMovie.overview}).props
      .children;

    expect(title).toBe(mockMovie.title);
    expect(releaseDate).toBe(mockMovie.release_date);
    expect(overview).toBe(mockMovie.overview);
  });
});
