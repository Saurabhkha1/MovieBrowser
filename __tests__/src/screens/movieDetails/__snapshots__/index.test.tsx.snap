// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails renders correctly 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#ccc",
        "borderRadius": 30,
        "height": 40,
        "justifyContent": "center",
        "left": 5,
        "opacity": 1,
        "position": "absolute",
        "top": 5,
        "width": 40,
        "zIndex": 1,
      }
    }
  >
    <MockedFontAwesomeIcon
      icon={
        {
          "icon": [
            320,
            512,
            [
              8249,
            ],
            "f104",
            "M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z",
          ],
          "iconName": "angle-left",
          "prefix": "fas",
        }
      }
      size={25}
    />
  </View>,
  <RCTScrollView
    style={
      {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
  >
    <View>
      <Image
        resizeMode="cover"
        source={
          {
            "uri": "https://image.tmdb.org/t/p/w500/test-poster.jpg",
          }
        }
        style={
          {
            "borderBottomColor": "#ccc",
            "borderBottomWidth": 1,
            "height": 500,
            "width": "100%",
          }
        }
      />
      <View
        style={
          {
            "padding": 15,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 24,
              "fontWeight": "bold",
              "marginBottom": 10,
            }
          }
        >
          Test Movie
        </Text>
        <Text
          style={
            {
              "color": "#999",
              "fontSize": 18,
              "marginBottom": 15,
            }
          }
        >
          2023-01-01
        </Text>
        <Text
          style={
            {
              "color": "#000",
              "fontSize": 16,
              "lineHeight": 24,
            }
          }
        >
          Test movie overview
        </Text>
      </View>
    </View>
  </RCTScrollView>,
]
`;
